assets/js/data.coffee

```diff
@@ -249,11 +263,67 @@ class Data
 
+  countInstances: (id, cache={}) ->
+    # Precondition: No circular references in ancestry
+    unless cache[id]?
+      errors.assert id?, "Empty id passed to countInstances"
+      if id == constants.root_id
+        return 1
+      parentCount = 0
+      for parent_id in (@store.getParents id)
+        parentCount += @countInstances parent_id, cache # Always exactly once under every parent
+      cache[id] = parentCount
+    cache[id]
+
+  exactlyOneInstance: (id) ->
+    1 == @countInstances id
+
   canonicalInstance: (id) -> # Given an id (for example with search or mark), return a row with that id
-    if id == @root.id
+    # TODO: Figure out which is the canonical one. Right now this is really 'arbitraryInstance'
+    # This probably isn't as performant as it could be for how often it gets called, but I'd rather make it called less often before optimizing.
+    errors.assert id?, "Empty id passed to canonicalInstance"
+    if id == constants.root_id
       return @root
-    canonicalParent = @canonicalInstance (@store.getParent id)
-    return @getChild canonicalParent, id
+    parentId = (@store.getParents id)[0] # This is the only actual choice made
+    errors.assert parentId?, "No parent found for id: #{id}"
+    canonicalParent = @canonicalInstance parentId
+    instance = @getChild canonicalParent, id
+    errors.assert instance?, "No canonical instance found for id: #{id}"
+    return instance
+
+  allAncestors: (id, options) ->
+    # Return all ancestor ids. Does not include <id>.
+    # If A is a parent of B, B is returned earlier in the list than A. This is called 'topological sort'.
+    options = _.defaults {}, options, { inclusive: false }
+    sorted = {}
+    preprocessed = {} # No repeats
+    ancestors = [] # Same contents as 'sorted' with preserved insert order
+    unless options.inclusive
+      sorted[id] = true # Leave out <id>
+    visit = (n) => # Do a DFS and add each node which points only to things the DFS has visited in turn (Tarjan's algorithm)
+      for parent in @store.getParents n
+        if parent not of preprocessed
+          preprocessed[parent] = true
+          visit parent
+      if n not of sorted
+        sorted[n] = true
+        ancestors.unshift n
+    visit id
+    ancestors
+
+  # Precondition: row.id == id
+  # Returns all ancestors of row, which are ancestors of only row and not any other instance of 'id'
+  # If A is a parent of B, B is returned earlier in the list than A. This is called 'topological sort'.
+  deltaAncestry: (id, row, options) ->
+    errors.assert_equals id, row.id, "Row is expected to be an instance of id"
+    options = _.defaults {}, options, { inclusive: false }
+
+    parentId = (do row.getParent).id
+    rowAncestry = @allAncestors parentId, { inclusive: options.inclusive }
+    idAncestry = []
+    for otherParentId in _.without (@getParents row), parentId
+      idAncestry = _.union idAncestry, (@allAncestors otherParentId, { inclusive: true })
```

Comment:
Nope. We're trying to figure which things are ancestors of id only via one parent, and not via any of the other parents--taking "via all parents" and subtracting "via this parent" does not work.